### PR TITLE
Unit tests added. Some improvement in training script.

### PR DESCRIPTION
**Pull request by Bozhi-Lyu**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,6 +17,7 @@ manual_test/
 
 # data file folder
 data/raw/fer2013.csv
+tests/testdata/*.pt
```
Comment (jakwisn) [inline .gitignore:20]: I think we need testdata for tests to run on github.

File: tests/test_CI.py [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
 
+
 def test_function():
     return np.sum(np.random.rand(5))
 
 
 if __name__ == "__main__":
-    test_function()
\ No newline at end of file
+    test_function()
```
Comment (jakwisn) [inline tests/test_CI.py:10]: Do we need that now?

File: tests/test_data.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    test_target = torch.load(os.path.join(_PATH_DATA, "processed", "test_target.pt"))
+
+    assert train_images.shape == torch.Size([28709, 1, 48, 48])
+    assert train_target.shape == torch.Size([28709])
+
+    assert validation_images.shape == torch.Size([3589, 1, 48, 48])
+    assert validation_target.shape == torch.Size([3589])
+
+    assert test_images.shape == torch.Size([3589, 1, 48, 48])
+    assert test_target.shape == torch.Size([3589])
```
Comment (jakwisn) [inline tests/test_data.py:23]: Does it use test data or normal data? Idea is not to use normal data in testing but the test data that you generated and that is available on github

File: tests/test_model.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import os
+import torch
+from tests import _PATH_MODEL
+from src.models.model import DeiTClassifier
+
+
+def test_model():
+    testmodel = DeiTClassifier()
+
+    model_path = os.path.join(_PATH_MODEL, "saved_models", "model.pt")
```
Comment (jakwisn) [inline tests/test_model.py:10]: We do not store model weights on github. Here we can initialize the model randomly - we are only looking at the shape not the actual output.

**Review by jakwisn — CHANGES_REQUESTED** [review]
Can you please remove test data file? Or add test data to data folder to see if dataloaders/transforms are working etc. Also removing model weights loading would be good.

File: tests/test_CI.py (hunk 2 above)
Comment (Bozhi-Lyu) [reply tests/test_CI.py]: No I don't think so. This file is not created by me at first and I just keep it.

File: src/predict_model.py [hunk 5]
```diff
@@ -5,52 +5,83 @@
...
 
+    pred_images = torch.load(os.path.join(config["data_path"] + "pred_images.pt"))
+    pred_target = torch.load(config["data_path"] + "pred_target.pt")
+    transform = transforms.Compose(
+        [
+            transforms.RandomHorizontalFlip(p=0.5),  # 50% chance of applying a horizontal flip
+            transforms.RandomRotation(10),  # Rotate the image by up to 10 degrees
+            transforms.RandomResizedCrop(48, scale=(0.8, 1.0)),  # Zoom in on the image
+        ]
+    )
```
Comment (jakwisn) [inline src/predict_model.py:47]: While doing prediction we dont want to transform images. It is only done on training

File: src/predict_model.py [hunk 6]
```diff
@@ -5,52 +5,83 @@
...
+
     model = DeiTClassifier().to(device)
     wandb.watch(model)
     model.load_state_dict(torch.load(model_checkpoint))
 
-    return torch.cat([model(batch) for batch in dataloader], 0)
+    trainloader = torch.utils.data.DataLoader(
+        pred_set,
+        batch_size=config.batch_size,
+        shuffle=True,
```
Comment (jakwisn) [inline src/predict_model.py:58]: shuffle = False - it is prediction

File: src/predict_model.py [hunk 7]
```diff
@@ -5,52 +5,83 @@
...
+    )
+
+    pred_set = CustomTensorDataset((pred_images, pred_target), transform=transform)
+
     model = DeiTClassifier().to(device)
     wandb.watch(model)
     model.load_state_dict(torch.load(model_checkpoint))
 
-    return torch.cat([model(batch) for batch in dataloader], 0)
+    trainloader = torch.utils.data.DataLoader(
```
Comment (jakwisn) [inline src/predict_model.py:55]: can we name it something else than `trainloader` it is not training

**Review by jakwisn — CHANGES_REQUESTED** [review]

File: src/data/make_subset.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        raw_dir (str): The directory where the raw data is stored.
+        processed_dir (str): The directory where processed data will be saved.
+        num(int): Number of samples for prediction.
+    """
+
+    logger.info("Processing dataset...")
+    logger.info(f"Raw directory: {raw_dir}")
+
+    # Reading dataset from CSV
+    df = pd.read_csv(raw_dir + "fer2013.csv")
```
Comment (jakwisn) [inline src/data/make_subset.py:31]: if we test with make subset we assume that csv is not on github
